cupy/core/core.pyx

```diff
@@ -1331,66 +1331,140 @@ divmod = create_ufunc(
 
 
 cdef _min_max_preamble = '''
+template <typename T>
 struct min_max_st{
-    type_in0_raw value;
+    T value;
     int index;
     __device__ min_max_st() : index(-1) { }
-    __device__ min_max_st(type_in0_raw v) : value(v), index(0) { }
-    __device__ min_max_st(type_in0_raw v, int i) : value(v), index(i) { }
+    __device__ min_max_st(T v) : value(v), index(0) { }
+    __device__ min_max_st(T v, int i) : value(v), index(i) { }
 };
-__device__ min_max_st my_min(const min_max_st& a, const min_max_st& b) {
+
+#define ISNAN(x) (x != x)
```

Comment:
This macro works for `float16` via [`operator float`](https://github.com/pfnet/chainer/blob/v1.13.0/cupy/core/carray.cuh#L35). It should be avoided to depend on such unobvious behavior. This behavior is an issue of cupy's float16, so it should be fixed in another PR.